api/envoy/config/route/v3/route_components.proto

```diff
@@ -760,6 +760,29 @@ message RouteAction {
     ConnectConfig connect_config = 3;
   }
 
+  message MaxStreamDuration {
+    // Specifies the maximum duration allowed for streams on the route. If not specified, the
+    // value from :ref:`max_stream_duration
+    // <envoy_api_field_config.core.v3.HttpProtocolOptions.max_stream_duration>` is used.
+    google.protobuf.Duration max_stream_duration = 1;
+
+    // If present, and the request is a gRPC request, use the
+    // `grpc-timeout header <https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md>`_,
+    // or its default value (infinity) as the max_stream_duration, but limit the applied timeout to
+    // the maximum value specified here. If configured as 0, the maximum allowed timeout for gRPC
+    // requests is infinity.
+    google.protobuf.Duration grpc_max_timeout = 2;
+
+    // If present, Envoy will adjust the timeout provided by the `grpc-timeout` header by
+    // subtracting the provided duration from the header. This is useful for allowing Envoy to set
+    // its global timeout to be less than that of the deadline imposed by the calling client, which
+    // makes it more likely that Envoy will handle the timeout instead of having the call canceled
+    // by the client. The offset will only be applied if the provided `grpc-timeout` header is
```

Comment:
The behavior of ignoring this when the timeout offset is greater than the timeout is interesting.  I see that that's what was done for the existing `grpc_timeout_offset` field, but is that actually what we want?  Internally, I believe that a lot of our LSTM infrastructure is specifically designed to proactively drop requests in this case, because if a server is overloaded and is suffering from high latency, we would generally rather fail a request that is not likely to succeed anyway, since allowing the request to proceed will not shed load and makes it harder for the server to recover.

Reply:
That makes more sense to me than the current behavior.  I was originally thinking we shouldn't change the semantics too much, to avoid surprises if someone simply uses the new fields instead of the old ones.  However, the semantics are already changing, so it seems okay.